### PR TITLE
Broadcast values to other processors to have a consistent GridFunction

### PR DESCRIPTION
**Pull request by jandrej**
See #4145 

Quality of life improvement

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4151](https://github.com/mfem/mfem/pull/4151) | @jandrej | @tzanio | @termi-official + @paul-hilscher | 2/25/24 | 3/29/24 | 3/31/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
We should also do a similar fix for internal boundaries coinciding with processor boundaries in `ProjectBdrCoefficientNormal()`, i.e. for H(div) spaces. Note that this method currently does not have a parallel version.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by termi-official** [comment]
Can we add a regression test for this?

**Comment by jandrej** [comment]
> Can we add a regression test for this?

We can. I can check on two ranks if the boundary transfer on a conforming mesh is the same between T-Vector and L-Vector. Is this a sufficient solution?

**Comment by najlkin** [comment]
As we have discussed at the meeting, I have done something like this already here #1951 :wink:.

**Comment by v-dobrev** [comment]
> As we have discussed at the meeting, I have done something like this already here #1951 😉.

Note that there's one unnecessary `Bcast` in #1951 compared to this.

On the other hand, #1951 correctly remove the comments
```
   // Only the values in the master are guaranteed to be correct!
```
from the modified methods.

**Comment by najlkin** [comment]
That `bcast` is needed because of the debugging check, where `values_counter` is compared with the essential vdofs, which are synchronized, if I am not wrong 🤔

**Comment by v-dobrev** [comment]
> That `bcast` is needed because of the debugging check, where `values_counter` is compared with the essential vdofs, which are synchronized, if I am not wrong 🤔

In the debug code, the check `pfes->GetLocalTDofNumber(i) == -1 || ...` ensures that the check is only done on the master processor for dof `i`.

**Comment by najlkin** [comment]
> > That `bcast` is needed because of the debugging check, where `values_counter` is compared with the essential vdofs, which are synchronized, if I am not wrong 🤔
> 
> In the debug code, the check `pfes->GetLocalTDofNumber(i) == -1 || ...` ensures that the check is only done on the master processor for dof `i`.

Hmm, you are right 😉 But it would actually make sense to check it everywhere when we broadcast the values now...

**Comment by najlkin** [comment]
Ok, to add some concurrency here 😁 , I added this check (removed `pfes->GetLocalTDofNumber(i) == -1`) to #1951 .

**Comment by v-dobrev** [comment]
> Ok, to add some concurrency here 😁 , I added this check (removed `pfes->GetLocalTDofNumber(i) == -1`) to #1951 .

That's exactly the opposite action to take after my comment.

In any case, this will be discussed separately in #1951, so let's not overtake this PR's discussion -- please, post any other comments related to #1951 on that PR's discussion.

**Comment by paul-hilscher** [comment]
@v-dobrev , 

I have lost a bit of a track here. Based on the discussion [here](https://github.com/mfem/mfem/pull/4151#issuecomment-1977599105) will we move forward with this PR here ?

@jandrej , 

any luck with bringing in a regression test ?

Thanks.

**Comment by tzanio** [comment]
Now that PR https://github.com/mfem/mfem/pull/3922 has been merged in `master`, please update this PR by:

1. merging `master` in here and resolving conflicts
2. replacing use of `double` with `real_t`
3. guarding double/single-only sections with `MFEM_USE_DOUBLE` and `MFEM_USE_SINGLE`
4. making sure tests run cleanly with default and `MFEM_PRECISION=single` configurations

See https://github.com/orgs/mfem/discussions/4207 for more details and/or ask @dylan-copeland for help.

File: fem/pgridfunc.cpp
Comment (v-dobrev) [inline fem/pgridfunc.cpp:695]: There are still some places like this where `double` has to be updated to `real_t`.

File: tests/unit/fem/test_project_bdr_par.cpp
Comment (v-dobrev) [inline tests/unit/fem/test_project_bdr_par.cpp:102]: To avoid this duplication, one can use `MPITypeMap<real_t>::mpi_type` instead of `MPI_DOUBLE`/`MPI_FLOAT`.

To simplify this a bit more we can define a macro like `MFEM_MPI_REAL_T`, hypre does this and defines `HYPRE_MPI_INT` and `HYPRE_MPI_BIG_INT`.

**Comment by tzanio** [comment]
Can I get some approvals please? We would like to include this bugfix in `mfem-4.7`.

File: fem/pgridfunc.cpp
Comment (paul-hilscher) [inline fem/pgridfunc.cpp:695]: minor: 
why
```
gcomm.Bcast<real_t>((*this).HostReadWrite());
```
instead of just
```
gcomm.Bcast<real_t>(HostReadWrite());
```
?

**Review by paul-hilscher — APPROVED** [review]
Thanks for this nice contribution. lgtm.

Would be good to address [this comment](https://github.com/mfem/mfem/pull/4151#discussion_r1542101656) before merging.

**Comment by tzanio** [comment]
thanks @termi-official and @paul-hilscher !

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: fem/pgridfunc.cpp
Comment (jandrej) [reply fem/pgridfunc.cpp:695]: Probably a me'ism. I like to make the use of "this" clear in the code I write.

File: tests/unit/fem/test_project_bdr_par.cpp
Comment (tzanio) [reply tests/unit/fem/test_project_bdr_par.cpp:102]: @v-dobrev, does https://github.com/mfem/mfem/pull/4172/commits/18cf9d7ea16981835b9036b3d72ebd91bd0d178c look OK?

Comment (v-dobrev) [reply tests/unit/fem/test_project_bdr_par.cpp:102]: Look good, thanks! I left a suggestion about a small typo.

**Pull request merged**
